File: Algorithms/Sorting/bubbleSort.py
import os
import random
import time

ARRAY_SIZE = 10
ELEM_LIMIT = 50
TIME_DELAY = 0.005


def initArrayValues() -> list:
    array = []

    for _ in range(ELEM_LIMIT):
        array.append(random.randrange(0, ARRAY_SIZE))

    return array


def barGraph(array: list) -> None:
    maxElement = max(array)

    for i in range(maxElement):
        for j in range(len(array)):
            if array[j] >= maxElement - i:
                print("*", end="")
            else:
                print(" ", end="")

        print()


def isSorted(array: list) -> bool:
    for i in range(len(array) - 1):
        if array[i] > array[i + 1]:
            return False

    return True


def bubbleSort(array: list) -> list:
    barGraph(array)

    for i in range(len(array) - 1):
        for j in range(len(array) - i - 1):
            if array[j] > array[j + 1]:
                array[j], array[j + 1] = array[j + 1], array[j]

                time.sleep(TIME_DELAY)
                os.system("cls")
                barGraph(array)

    return array


def main() -> None:
    array = initArrayValues()
    array = bubbleSort(array)

    while True:
        if isSorted(array):
            print("DONE")
            break
        else:
            array = bubbleSort(array)


if __name__ == "__main__":
    main()

File: Algorithms/Sorting/bubbleSort.js
var ARRAY_SIZE = 50;
var ELEM_LIMIT = 10;
var TIME_DELAY = 0.005;
function initArrayValues() {
    return Array.apply(null, Array(ARRAY_SIZE)).map(function () {
        return Math.round(Math.random() * ELEM_LIMIT);
    });
}
function sleep(ms) {
    return new Promise(function (resolve) { return setTimeout(resolve, ms); });
}
function barGraph(array) {
    var maxElement = Math.max.apply(Math, array);
    for (var i = 0; i < maxElement; i++) {
        for (var j = 0; j < array.length; j++) {
            if (array[j] >= maxElement - i) {
                process.stdout.write("*");
            }
            else {
                process.stdout.write(" ");
            }
        }
        console.log();
    }
}
function isSorted(array) {
    for (var i = 0; i < array.length - 1; i++) {
        if (array[i] > array[i + 1]) {
            return false;
        }
    }
    return true;
}
function bubbleSort(array) {
    var _a;
    barGraph(array);
    for (var i = 0; i < array.length - 1; i++) {
        for (var j = 0; j < array.length - 1 - i; j++) {
            if (array[j] > array[j + 1]) {
                _a = [array[j + 1], array[j]], array[j] = _a[0], array[j + 1] = _a[1];
                sleep(TIME_DELAY);
                console.clear();
                barGraph(array);
            }
        }
    }
    return array;
}
function main() {
    var array = initArrayValues();
    array = bubbleSort(array);
    while (true) {
        if (isSorted(array)) {
            console.log("DONE");
            break;
        }
        else {
            array = bubbleSort(array);
        }
    }
    console.table(array);
    barGraph(array);
}
main();
